src/accelerator.rs

```diff
@@ -0,0 +1,467 @@
+//! Accelerators describe keyboard shortcuts defined by the application.
+//!
+//! [`Accelerator`s](crate::accelerator::Accelerator) are used to define a keyboard shortcut consisting
+//! of an optional combination of modifier keys (provided by [`SysMods`](crate::accelerator::SysMods),
+//! [`RawMods`](crate::accelerator::RawMods) or [`Modifiers`](crate::accelerator::Modifiers)) and
+//! one key ([`Code`](crate::accelerator::Code)).
+//!
+//! # Examples
+//! They can be created directly
+//! ```
+//! # use tao::accelerator::{Accelerator, AcceleratorId, SysMods, RawMods};
+//! # use tao::keyboard::{Modifiers, Code};
+//! #
+//! let accelerator = Accelerator::new(SysMods::Shift, Code::KeyQ);
+//! let accelerator_with_raw_mods = Accelerator::new(RawMods::Shift, Code::KeyQ);
+//! let accelerator_without_mods = Accelerator::new(None, Code::KeyQ);
+//! # assert_eq!(accelerator, accelerator_with_raw_mods);
+//! # assert_eq!(
+//! #     accelerator.with_id(AcceleratorId::new("shift+q")),
+//! #     "shift+q".parse().unwrap()
+//! # );
+//! ```
+//! or from `&str`, note that all modifiers
+//! have to be listed before the non-modifier key, `shift+alt+q` is legal,
+//! whereas `shift+q+alt` is not.
+//! ```
+//! # use tao::accelerator::{Accelerator, AcceleratorId, RawMods};
+//! # use tao::keyboard::{Modifiers, Code};
+//! #
+//! let accelerator: Accelerator = "shift+alt+q".parse().unwrap();
+//! # assert_eq!(
+//! #     accelerator,
+//! #     Accelerator::new(RawMods::AltShift, Code::KeyQ)
+//! #         .with_id(AcceleratorId::new("shift+alt+q"))
+//! # );
+//! #
+//! # // This assert exists to ensure a test breaks once the
+//! # // statement above about ordering is no longer valid.
+//! # assert!("shift+q+alt".parse::<Accelerator>().is_err());
+//! ```
+//!
+
+pub use keyboard_types::{Code, Modifiers};
+use std::{
+    borrow::Borrow,
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+    str::FromStr,
+};
+
+/// Base `Accelerator` functions.
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub struct Accelerator {
+    id: Option<AcceleratorId>,
+    pub(crate) mods: Modifiers,
+    pub(crate) key: Code,
+}
+
+impl Accelerator {
+    /// Creates a new accelerator to define keyboard shortcuts throughout your application.
+    pub fn new(mods: impl Into<Option<Modifiers>>, key: Code) -> Self {
+        Self {
+            id: None,
+            mods: mods.into().unwrap_or_else(Modifiers::empty),
+            key,
+        }
+    }
+
+    /// Assign a custom accelerator id.
+    pub fn with_id(mut self, id: AcceleratorId) -> Self {
+        self.id = Some(id);
+        self
+    }
+
+    /// Returns an identifier unique to the accelerator.
+    pub fn id(self) -> AcceleratorId {
+        if let Some(id) = self.id {
+            return id;
+        }
+
+        AcceleratorId(hash_accelerator_to_u16(self))
+    }
```

Comment:
Not needed. Accelerator is kind of a metadata, they serve nothing on their own. They don't need ids.